src/check/parse/Parser.zig

```diff
@@ -878,6 +945,244 @@ pub fn parseBody(self: *Parser) IR.NodeStore.BodyIdx {
     }
 }
 
+pub fn parseTypeHeader(self: *Parser) IR.NodeStore.TypeHeaderIdx {
+    const start = self.pos;
+    std.debug.assert(self.peek() == .UpperIdent);
+    self.advance(); // Advance past UpperIdent
+    if (self.peek() != .LowerIdent) {
+        return self.store.addTypeHeader(.{
+            .name = start,
+            .args = &.{},
+            .region = .{ .start = start, .end = start },
+        });
+    }
+    const scratch_top = self.store.scratch_tokens.items.len;
+    defer self.store.scratch_tokens.shrinkRetainingCapacity(scratch_top);
+    while (self.peek() == .LowerIdent) {
+        self.store.scratch_tokens.append(self.pos) catch exitOnOom();
+        self.advance(); // Advance past LowerIdent
+    }
+    const args = self.store.scratch_tokens.items[scratch_top..];
+    return self.store.addTypeHeader(.{
+        .name = start,
+        .args = args,
+        .region = .{ .start = start, .end = self.pos },
+    });
+}
+
+const TyFnArgs = enum {
+    not_looking_for_args,
+    looking_for_args,
+};
+
+pub fn parseTypeAnno(self: *Parser, looking_for_args: TyFnArgs) IR.NodeStore.TypeAnnoIdx {
+    const start = self.pos;
+    var anno: ?IR.NodeStore.TypeAnnoIdx = null;
+
+    switch (self.peek()) {
+        .UpperIdent => {
+            if (self.peekNext() != .NoSpaceOpenRound) {
+                anno = self.store.addTypeAnno(.{ .tag = .{
+                    .tok = self.pos,
+                    .args = &.{},
+                    .region = .{ .start = start, .end = self.pos },
+                } });
+                self.advance(); // Advance past UpperIdent
+            } else {
+                self.advance(); // Advance past UpperIdent
+                self.advance(); // Advance past NoSpaceOpenRound
+                const scratch_top = self.store.scratch_type_annos.items.len;
+                defer self.store.scratch_type_annos.shrinkRetainingCapacity(scratch_top);
+                // The following comparison should be made against all the tokens matched upon
+                // in the parent switch.
+                while (self.peek() != .CloseRound) {
+                    self.store.scratch_type_annos.append(self.parseTypeAnno(.looking_for_args)) catch exitOnOom();
+                    if (self.peek() != .Comma) {
+                        break;
+                    }
+                    self.advance(); // Advance past Comma
```

Comment:
Random note - in my recursive descent parsers, I. often include a utility method for checking whether the next token is a specific token and if so advancing - I think adding that could make quite a lot of this parser simpler.

Reply:
That's a great idea - I usually do too, I usually call it `accept`